Src/HALAL/Services/Time/Scheduler.cpp

```diff
@@ -0,0 +1,308 @@
+/*
+ * Scheduler.hpp
+ *
+ * Created on: 17 nov. 2025
+ *     Author: Victor (coauthor Stephan)
+ */
+#include "HALAL/Services/Time/Scheduler.hpp"
+
+// This is needed to register a TimerPeripheral
+#include "HALAL/Models/TimerPeripheral/TimerPeripheral.hpp"
+
+#include <algorithm>
+#include <limits>
+
+/* NOTE(vic): Pido perdón a Boris pero es la mejor manera que se me ha ocurrido hacer esto */
+#define SCHEDULER_RCC_TIMER_ENABLE \
+    glue(glue(RCC_APB1LENR_TIM, SCHEDULER_TIMER_IDX), EN)
+#define SCHEDULER_GLOBAL_TIMER_IRQn \
+    glue(TIM, glue(SCHEDULER_TIMER_IDX, _IRQn))
+#define SCHEDULER_GLOBAL_TIMER_CALLBACK() \
+    extern "C" void glue(TIM, glue(SCHEDULER_TIMER_IDX, _IRQHandler))(void)
+
+#define Scheduler_global_timer ((TIM_TypeDef*)Scheduler::global_timer_base)
+
+/* bit field insert
+ * (internally) mask = ((1UL << width) - 1) << pos
+ * dest = (dest & ~mask) | (source & mask)
+ * With the condition that pos and width are immediate values
+ * ref: https://developer.arm.com/documentation/dui0646/c/The-Cortex-M7-Instruction-Set/Bit-field-instructions/BFC-and-BFI
+ */
+#define BFI(dest, source, pos, width) \
+    __asm__( \
+        "bfi %0, %1, %2, %3" \
+        : "+r" (dest) \
+        : "r" (source), "i" (pos), "i" (width) \
+    )
+
+namespace {
+constexpr uint64_t kMaxIntervalUs =
+    static_cast<uint64_t>(std::numeric_limits<uint32_t>::max()) + 1ULL;
+}
+
+std::array<Scheduler::Task, Scheduler::kMaxTasks> Scheduler::tasks_{};
+uint64_t Scheduler::sorted_task_ids_ = 0;
+std::size_t Scheduler::active_task_count_{0};
+
+uint32_t Scheduler::ready_bitmap_{0};
+uint32_t Scheduler::used_bitmap_{0};
+uint64_t Scheduler::global_tick_us_{0};
+uint64_t Scheduler::current_interval_us_{0};
+
+inline uint8_t Scheduler::get_at(std::size_t idx) {
+    uint32_t shift = idx*4;
+    return (sorted_task_ids_ & (0x0F << shift)) >> shift;
+}
+inline void Scheduler::set_at(std::size_t idx, uint8_t id) {
+    uint32_t shift = idx*4;
+    uint64_t clearmask = ~(0xFF << shift);
+    Scheduler::sorted_task_ids_ = (sorted_task_ids_ & clearmask) | (id << shift);
+    // sorted_task_ids_ |= ((id & 0x0F) << shift); // This is also an option in case id is incorrect, I don't think it's necessary though
+}
+inline uint8_t Scheduler::front_id() {
+    return Scheduler::get_at(0);

```

Comment:
for front_id I would not use the genertic get at, there is a fastpath if you know it is 0;
return ((uint32_t*)sorted_task_list)[0] & 0xF